### PR TITLE
Layer refactor

### PR DESCRIPTION
**Pull request by SergioMartin86**
+ Added RNN
+ Refactored layer code. It's more modular now
+ Refactored NN definition on examples so one simply define hidden layers
+ Refactored supervised learning exp definition in RL methods
+ Added time sequence support on NN
+ Improved checkpoint/resume logic
+ Removed unused RL algos
+ Changed RL algos to use timesequence (instead of individual states)
+ Added REFER as common tool for all algorithms to use
+ Milestone: GFPT solving ABF2D in 7 mins
+ Milestone: VRACER solving ABF3D
+ Added Retrace as common tool for all algorithms
+ Improved statistics

### CODE REVIEW

File: examples/learning/supervised/run-rnn.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+### Using a neural network solver (deep learning) for inference
+
+e["Solver"]["Type"] = "Learner/DeepSupervisor"
+e["Solver"]["Loss Function"] = "Mean Squared Error"
+e["Solver"]["Steps Per Generation"] = 20
+e["Solver"]["Optimizer"] = "AdaBelief"
+e["Solver"]["Learning Rate"] = 0.0001
+
+### Defining the shape of the neural network
+
```
Comment (wadaniel) [inline examples/learning/supervised/run-rnn.py:69]: P please check if this configuration makes sense to you, it looks good to me

**Review by wadaniel — DISMISSED** [review]

File: examples/learning/reinforcement/cartpole/run-vracer-discrete.py [hunk 2]
```diff
@@ -34,45 +34,46 @@
...
 
-e["Solver"]["Critic"]["Neural Network"]["Layers"][2]["Type"] = "Layer/Dense"
-e["Solver"]["Critic"]["Neural Network"]["Layers"][2]["Node Count"] = 32
-e["Solver"]["Critic"]["Neural Network"]["Layers"][2]["Activation Function"]["Type"] = "Elementwise/Tanh"
+e["Solver"]["Neural Network"]["Hidden Layers"][2]["Type"] = "Layer/Linear"
+e["Solver"]["Neural Network"]["Hidden Layers"][2]["Output Channels"] = 32
 
-e["Solver"]["Critic"]["Neural Network"]["Layers"][3]["Type"] = "Layer/Dense"
-e["Solver"]["Critic"]["Neural Network"]["Layers"][3]["Activation Function"]["Type"] = "Elementwise/Linear"
+e["Solver"]["Neural Network"]["Hidden Layers"][3]["Type"] = "Layer/Activation"
```
Comment (webepasc) [inline examples/learning/reinforcement/cartpole/run-vracer-discrete.py:75]: I must say I find this confusing. Technically applying an activation function is not a layer...

File: examples/learning/supervised/run-rnn.py (hunk 1 above)
Comment (webepasc) [reply examples/learning/supervised/run-rnn.py:69]: I think we have to discuss this...

File: source/modules/neuralNetwork/layer/recurrent/README.rst [hunk 3]
```diff
@@ -0,0 +1,5 @@
+*************
+Recurrent
+*************
+
+Represent a hidden layer where a recurrent transformation is applied.
```
Comment (webepasc) [inline source/modules/neuralNetwork/layer/recurrent/README.rst:5]: We should add a proper documentation..

File: source/modules/neuralNetwork/layer/recurrent/gru/README.rst [hunk 4]
```diff
@@ -0,0 +1,5 @@
+*************
+GRU
+*************
+
+Uses the Gated recurrent units (GRUs) architecture.
```
Comment (webepasc) [inline source/modules/neuralNetwork/layer/recurrent/gru/README.rst:5]: We should add a proper documentation..

**Review by webepasc — APPROVED** [review]
There is a lot of changes in there. I dislike some of the changes to layer. I think we should discuss it. Also, I am still not too happy about the removal of DDPG, ACER, and dueling DQN. I would rather prefer to push forward the automatic testing for the modules rather than removing them...

**Comment by wadaniel** [comment]
I noticed that the naf cartpole experiment segfaults and vracer cartpole is not converging

**Comment by SergioMartin86** [comment]
> I noticed that the naf cartpole experiment segfaults and vracer cartpole is not converging

Fixed them both. The segfault was a wrong access to the circular buffer (the access was wrong, but there was a bug in cBuffer that made it not fail. After I fixed the bug in cBuffer, the wrong access segfaulted. They are both fixed now.)

As per convergence, there was a copy-paste mistake in both NAF and VRACER where the supervisor was using "MSE" instead of "Direct Gradients". Should be fixed now.

**Review by wadaniel — DISMISSED** [review]

**Review by webepasc — APPROVED** [review]

**Review by webepasc — APPROVED** [review]

**Review by wadaniel — APPROVED** [review]

**Pull request merged**